src/tunnel.cc

```diff
@@ -995,38 +1047,89 @@ tunnelStart(ClientHttpRequest * http)
+        if (p->secure.encryptTransport)
+            return advanceDestination("secure connection to peer", conn, [this,&conn] {
+                secureConnectionToPeer(conn);
+            });
+    }
+
+    connectedToPeer(conn);
+}
+
+/// encrypts an established TCP connection to peer
+void
+TunnelStateData::secureConnectionToPeer(const Comm::ConnectionPointer &conn)
+{
+    { // TODO: Remove this diff reduction before commit
+        { // TODO: Remove this diff reduction before commit
             AsyncCall::Pointer callback = asyncCall(5,4,
                                                     "TunnelStateData::ConnectedToPeer",
-                                                    MyAnswerDialer(&TunnelStateData::connectedToPeer, this));
-            auto *connector = new Security::BlindPeerConnector(request, server.conn, callback, al);
+                                                    MyAnswerDialer(&TunnelStateData::noteSecurityPeerConnectorAnswer, this));
+            const auto connector = new Security::BlindPeerConnector(request, conn, callback, al);
             AsyncJob::Start(connector); // will call our callback
-            return;
         }
     }
+}
 
-    Security::EncryptorAnswer nil;
-    connectedToPeer(nil);
+/// starts a preparation step for an established connection; retries on failures
+template <typename StepStart>
+void
+TunnelStateData::advanceDestination(const char *stepDescription, const Comm::ConnectionPointer &conn, const StepStart &startStep)
+{
+    // TODO: Extract destination-specific handling from TunnelStateData so that
+    // all the awkward, limited-scope advanceDestination() calls can be replaced
+    // with a single simple try/catch,retry block.
+    try {
+        startStep();
+        // now wait for the step callback
+    } catch (...) {
+        debugs (26, 2, "exception while trying to " << stepDescription << ": " << CurrentException);
+        closePendingConnection(conn, "connection preparation exception");
+        retryOrBail();
+    }
 }
 
+/// callback handler for the connection encryptor
 void
-TunnelStateData::connectedToPeer(Security::EncryptorAnswer &answer)
+TunnelStateData::noteSecurityPeerConnectorAnswer(Security::EncryptorAnswer &answer)
 {
     if (ErrorState *error = answer.error.get()) {
         answer.error.clear(); // sendError() will own the error
         sendError(error, "TLS peer connection error");
         return;
     }
 
+    if (!Comm::IsConnOpen(answer.conn) || fd_table[answer.conn->fd].closing()) {
+        sendError(new ErrorState(ERR_CANNOT_FORWARD, Http::scServiceUnavailable, request.getRaw(), al), "connecion gone");
+        return;
+    }
+
+    connectedToPeer(answer.conn);
+}
+
+void
+TunnelStateData::connectedToPeer(const Comm::ConnectionPointer &conn)
+{
+    advanceDestination("establish tunnel thru proxy", conn, [this,&conn] {
+        establishTunnelThruProxy(conn);
+    });
+    // XXX: In tests, if establishTunnelThruProxy() throws and
```

Comment:
@chtsanti, please try to reproduce this XXX. Why do we not get the expected outcome here?

Reply:
In my tests I am seeing that this is not correct, the message will appear after the last owner of the connection object, the AccessLogEntry::hier::tcpServer, be destroyed.
I removed this XXX comment in my commit 6eab3c8.

However while working for this I found other problems with one crash fixed in commit b5d3dcf enough serious. I added also some other fixes please review.